### ISSUE TITLE
ConnectionResetError: [Errno 104] Connection reset by peer

### ISSUE THREAD

**Issue by d1618033** [ronf/asyncssh#259]
I get the above error when running the following code:

```python
await asyncio.wait_for(
    asyncssh.connect(
        hostname,
        password=...,
        username=...,
        known_hosts=None,
        client_keys=[....],
    ),
    timeout=5,
)
```
The traceback shows that the problem originated with peername being None
```
Exception in callback SSHConnection.connection_made(<_SelectorSoc...e, bufsize=0>>)
handle: <Handle SSHConnection.connection_made(<_SelectorSoc...e, bufsize=0>>)>
Traceback (most recent call last):
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File ".../lib/python3.8/site-packages/asyncssh/connection.py", line 692, in connection_made
    self._peer_addr, self._peer_port = peername[:2]
TypeError: 'NoneType' object is not subscriptable
```

I found [this](https://github.com/ronf/asyncssh/issues/16#issuecomment-106666655) issue where you explained that peername can be None and that it's best to check if it's None before attempting to split it.
So is it as a simple as checking if peername is None and if it is skipping over that line? If so, I'd be happy to make a PR with a fix.

**Comment by ronf** [comment]
I think there's something more complicated going on here. If the timeout actually fired and triggered an exception before the socket successfully connected, I would not have expected the connection_made callback to be called. That's why the assignment there is unconditional -- it should only be called when a connection is successfully opened, and in that case, peername should have been set to something.

I'll see if I can reproduce this and dig a bit more this weekend into how peername is becoming None. Thanks for the report!

**Comment by ronf** [comment]
In a quick test here, I was not able to reproduce this. I got the expected TimeoutError:

```
INFO:asyncssh:Opening SSH connection to 1.1.1.1, port 22
Traceback (most recent call last):
  File "./simple_client.py", line 49, in <module>
    loop.run_until_complete(asyncio.wait_for(run_client(), timeout=1))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/base_events.py", line 612, in run_until_complete
    return future.result()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/tasks.py", line 490, in wait_for
    raise exceptions.TimeoutError()
asyncio.exceptions.TimeoutError
```

I get the same result when wrapping the asyncssh.connect() call directly with asyncio.wait_for(), as you did in your example here.

**Comment by d1618033** [comment]
Thank you for looking into this.
I should have mentioned that the error doesn't happen in a consistent way - it works most of the time. It's just that some of the time peername returns None and then I get the connection reset by peer error. 
I will try to work on making a code snippet that reproduces this problem in a consistent way

**Comment by ronf** [comment]
It's interesting that you're getting an 'Connection reset by peer' error here, rather than a 'Connection refused' or timeout. That implies the remote system did accept the connection, but then ended up resetting it after that, but quickly enough that perhaps the client didn't have time to determine the peer information from the connection before it completed the TCP handshake.

I guess there shouldn't be any harm in coding defensively for this, as you proposed. I'd just like to understand a bit better how this is happening, in case other changes might be needed as well.

**Comment by d1618033** [comment]
Your theory makes sense. Is there a way to test it? I tried to set up something in the unittests but got lost in the details.

I investigated a bit why peername is None and found that it happens in the asyncio package and unfortunately, they hide the exception raised by `socket.getpeername` so I'm having trouble figuring out why it happened. 

```python
class _SelectorTransport(transports._FlowControlMixin,
                         transports.Transport):

    max_size = 256 * 1024  # Buffer size passed to recv().

    _buffer_factory = bytearray  # Constructs initial value for self._buffer.

    # Attribute used in the destructor: it must be set even if the constructor
    # is not called (see _SelectorSslTransport which may start by raising an
    # exception)
    _sock = None

    def __init__(self, loop, sock, protocol, extra=None, server=None):
        super().__init__(extra, loop)
        self._extra['socket'] = trsock.TransportSocket(sock)
        try:
            self._extra['sockname'] = sock.getsockname()
        except OSError:
            self._extra['sockname'] = None
        if 'peername' not in self._extra:
            try:
                self._extra['peername'] = sock.getpeername()
            except socket.error:
                self._extra['peername'] = None
```

I opened an issue at bugs.python.org (https://bugs.python.org/issue39700), but they have yet to respond :(

**Comment by d1618033** [comment]
I don't know if this helps, but I looked in the logs in the host and found this:

```
sshd: Bad protocol version identification '' from <ip> port <port>
```
Any idea what could cause this?

**Comment by ronf** [comment]
I think that particular "Bad protocol version identification" happens any time the socket is closed without sending the initial SSH version string. So, if the socket closes for some reason before the application can write to it (which would be consistent with the getpeername() failing), you'd see that error.

The fact that they are catching and ignoring the exception when they set the 'peername' value to None could explain why connection_made() is still being called. The code snippet you posted here is definitely a good reason to add the defensive code which checks the value before unpacking it.

As for how this could happen, I'm still thinking it might be related to the TCP connection failing for some reason after it connects but before the _SelectorTransport() is initialized. I don't know what would cause that, and it seems like it would be a pretty small window of time, but if the connection was already broken before the call to sock.getpeername(), I think we'd see what you have described here. You might want to add a debugging statement to print out the details of the socket.error exception in the code you included here, though I suspect that will just confirm that it is a "Connection reset by peer" error and not explain why that's happening.

You might also want to do a packet capture at the TCP level to see if that shows anything interesting. At the very least, that'll let you see which side the connection reset is coming from.

**Comment by d1618033** [comment]
Thank you. 

I'm not getting the error any more, but if it starts happening again I will try your packet capture suggestion.

In the meantime, is there anything you want me to change with my pull request? e.g, maybe if peername is None an exception should be raised?

**Comment by ronf** [comment]
I'll probably bring this change into the "develop" branch first. I appreciate the pull request, but I'll probably commit something separate there along the same lines once I'm done with some other changes.

I'm thinking this protection should probably be added for both the sockname and the peername, and I'll either need to add in some unit tests or add a #pragma: no branch to make coverage happy on the case where either of those actually is `None`.

**Comment by ronf** [comment]
The defensive code to handle sockname/peername being None is now commit a7e5bfb in the "develop" branch. Thanks for identifying this issue and the related code in asyncio where it can happen, as well as your proposed fix! This'll go out in a new release shortly.

**Comment by ronf** [comment]
This change is now available in AsyncSSH 2.2.0.
